software/py/blood_graph.py

```diff
@@ -3,48 +3,246 @@
+                          # but treated the same as stall_ifetch_wait
+                          #"icache_miss",
+                          "lr",
+                          "lr_aq",
+                          "swap_aq",
+                          "swap_rl",
+                          "beq",
+                          "bne",
+                          "blt",
+                          "bge",
+                          "bltu",
+                          "bgeu",
+                          "jalr",
+                          "jal",
+                          "beq_miss",
+                          "bne_miss",
+                          "blt_miss",
+                          "bge_miss",
+                          "bltu_miss",
+                          "bgeu_miss",
+                          "jalr_miss",
+                          "sll",
+                          "slli",
+                          "srl",
+                          "srli",
+                          "sra",
+                          "srai",
+                          "add",
+                          "addi",
+                          "sub",
+                          "lui",
+                          "auipc",
+                          "xor",
+                          "xori",
+                          "or",
+                          "ori",
+                          "and",
+                          "andi",
+                          "slt",
+                          "slti",
+                          "sltu",
+                          "sltiu",
+                          "mul",
+                          "mulh",
+                          "mulhsu",
+                          "mulhu",
+                          "div",
+                          "divu",
+                          "rem",
+                          "remu",
+                          "fence" }
+
+
+    # List of types of floating point instructions executed by the core
+    self.fp_instr_list = {"fadd",
+                          "fsub",
+                          "fmul",
+                          "fsgnj",
+                          "fsgnjn",
+                          "fsgnjx",
+                          "fmin",
+                          "fmax",
+                          "fcvt_s_w",
+                          "fcvt_s_wu",
+                          "fmv_w_x",
+                          "feq",
+                          "flt",
+                          "fle",
+                          "fcvt_w_s",
+                          "fcvt_wu_s",
+                          "fclass",
+                          "fmv_x_w" }
+
+    # List of unknown operation by the core 
+    self.unknown_list  = {"unknown"}
+
+
+    # Coloring scheme for different types of operations
+    # For detailed mode 
+    # i_cache miss is treated the same is stall_ifetch_wait
+    self.detailed_stall_bubble_color = { "stall_depend"                           : (0x00, 0x00, 0x80), ## navy blue
```

Comment:
A thought - but hex colors aren't very intuitive. What if we used the Python Colour package (https://github.com/vaab/colour) or the ImageColor package from PIL (https://pillow.readthedocs.io/en/3.1.x/reference/ImageColor.html#color-names), and use the color names defined there. 

That would allow something like: 

```
stall_depend :       Color("NavyBlue")
stall_depend_local_load: Color("Cyan")
```
 or

```
stall_depend :       ImageColor.getrgb("NavyBlue")
stall_depend_local_load: ImageColor.getrgb("Cyan")
```

Reply:
Yeah in general it is more graspable, and I thought about using coloring names at first too. But the problem is there are too many colors, I needed between about 16 highly distinguishable colors and it was too hard to pick those using just your eye. So Michael came up with the idea of dividing the range of 0x00 to 0xff for each of the R,G,B colors into however many shades of that color we want, and generate the combination of those colors. I've done the same in the bloodgraph more or less, or at least tried to come close to it. I've used two shades of Red and Blue (0x00 & 0xff) and four shades of Green (0x00, 0x55, 0xaa, 0xff) to generate the colors. So the colors don't necessarily have a name, and I need to fiddle with the hex code too often to generate better colors so I'd rather it stay as hex coding. Why four shades of green you might wonder, because human eye can distinguish between more shades of Green than any other color :D